'use strict';
/**
 * Converts given value to BN object if it is number or string. Otherwise defaultValue is
 * returned in case given value is not truthy.
 *
 * @param {number|string|BN|null} number
 * @param {number|string|BN|null} [defaultValue]
 * @returns {BN|null}
 */

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bn = bn;
exports.expectRevert = expectRevert;
exports.expectRevertOrFail = expectRevertOrFail;

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function bn(number) {
  var defaultValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

  if (number == null) {
    if (defaultValue == null) {
      return null;
    }

    number = defaultValue;
  }

  return web3.utils.toBN(number);
}
/**
 * Asserts that given promise will throw because of revert().
 * @param {Promise} promise
 */


function expectRevert(_x) {
  return _expectRevert.apply(this, arguments);
}
/**
 * Asserts that given promise will throw because of revert() or failed assertion.
 * @param {Promise} promise
 */


function _expectRevert() {
  _expectRevert = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(promise) {
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return expectError(promise, ['revert']);

          case 2:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _expectRevert.apply(this, arguments);
}

function expectRevertOrFail(_x2) {
  return _expectRevertOrFail.apply(this, arguments);
}
/**
 * Asserts that given promise will throw and that thrown message will contain one of the given
 * search strings.
 *
 * @param {Promise} promise The promise expecting to throw.
 * @param {string[]} messages List of expected thrown message search strings.
 */


function _expectRevertOrFail() {
  _expectRevertOrFail = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee2(promise) {
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return expectError(promise, ['revert', 'invalid opcode']);

          case 2:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _expectRevertOrFail.apply(this, arguments);
}

function expectError(_x3, _x4) {
  return _expectError.apply(this, arguments);
}

function _expectError() {
  _expectError = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee3(promise, messages) {
    var i;
    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.prev = 0;
            _context3.next = 3;
            return promise;

          case 3:
            _context3.next = 15;
            break;

          case 5:
            _context3.prev = 5;
            _context3.t0 = _context3["catch"](0);
            i = 0;

          case 8:
            if (!(i < messages.length)) {
              _context3.next = 14;
              break;
            }

            if (!(_context3.t0.message.search(messages[i]) >= 0)) {
              _context3.next = 11;
              break;
            }

            return _context3.abrupt("return");

          case 11:
            i++;
            _context3.next = 8;
            break;

          case 14:
            assert.fail("Expected revert, got '" + _context3.t0 + "' instead.");

          case 15:
            assert.fail('Expected revert not received.');

          case 16:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3, null, [[0, 5]]);
  }));
  return _expectError.apply(this, arguments);
}